### ISSUE TITLE
How should we handle wrapped errors?

### ISSUE THREAD

**Issue by JustinBeckwith** [googleapis/google-cloud-node#2912]
In `google-auth-library`, we end up with a fair number of bugs like this one:
https://github.com/googleapis/google-auth-library-nodejs/issues/643

The TL;DR is that this pattern is not uncommon:

```js
    let isGCE;
    try {
      isGCE = await this._checkIsGCE();
    } catch (e) {
      throw new Error('Unexpected error determining execution environment: ' + e.message);
    }
```

*note: I didn't originally write this code.*

This is bad.  It's bad because it hides the original error from the user, which makes it impossible for them to know what actually went wrong.  I see the options here as:
1. We just let the original error throw with no additional context
2. We create a wrapped error class that exposes an `originalError` property 

Perhaps orthogonal, there's also the dreaded issue of coded errors, and inner error parsing:
https://github.com/googleapis/nodejs-common/blob/master/src/util.ts#L119

With this logic wrapped up in a fairly large common module, it's hard to use that thing throughout all of our modules.  If we think this wrapped error thing is worth having... it may make sense to create a `@google-cloud/error` package that has all of our particulars in it.  

What are folks thoughts here?  I'm just kinda throwing out problems, and I want to see what bounces back :)

@googleapis/node-team @googleapis/yoshi-nodejs

**Comment by callmehiphop** [comment]
I ran into a very similar issue in PubSub. My strategy was to subclass errors and prepend some additional verbiage to the original error message ([example](https://github.com/googleapis/nodejs-pubsub/blob/master/src/message-stream.ts#L106)), it's only been a couple months but it *appears* to have helped quite a bit.

**Comment by sio4** [comment]
First, APIs are the face of Google Cloud products for developer customers, and it is important that how is it easier to solve problems. If we improved the messages, it will increase the supportability of products. Please consider about this.

For the topic, if we just add message prefix looks like "Something went wrong...", then we don't need to catch it and throw it again with meaningless prefix. But, If we add a little message which is descriptive, it will help us a lot.
In my opinion, it is enough that we just add that kind of contextual messages within routines in the module since the context depends on each module and situation. It is hard to make/manage common error module as fit to most of the modules. Again, meaning less prefix is not necessary and meaningless passing the original exception message also not necessary.

**Comment by bcoe** [comment]
In yargs, similarly to @callmehiphop, we ended up sub-classing Error:

https://github.com/yargs/yargs/blob/master/lib/yerror.js

This gave us more flexibility to capture errors, and attach more contextual information, while not capturing errors produced externally from the yargs library -- I'd like this approach.

**Comment by JustinBeckwith** [comment]
For places where I ran into this, we've updated it to append text to the message, and then re-throw the original exception to keep the stacktrace in order 👍
